### PR TITLE
CP-10531: Update vm_guest_metrics to implement PV drivers update

### PR DESCRIPTION
**Pull request by huizh**
1. Introduce networks_path_optimized and storages_path_optimized flags to vm_guest_metrics.
2. Update the implementation of PV_drivers_up_to_date.

Signed-off-by: Hui Zhang <hui.zhang@citrix.com>

### CODE REVIEW

**Comment by cheng--zhang** [comment]
Hi @simonjbeaumont 
Could you please help to review the code?
Thanks

**Comment by huizh** [comment]
FYI.
It's originally reviewed on master branch.
https://github.com/xapi-project/xen-api/pull/2160
Thanks.

File: ocaml/idl/datamodel.ml [hunk 1]
```diff
@@ -7050,9 +7050,21 @@ let vm_guest_metrics =
       field ~qualifier:DynamicRO ~ty:(Map(String, String)) "os_version" "version of the OS";
       field ~qualifier:DynamicRO ~ty:(Map(String, String)) "PV_drivers_version" 
 	"version of the PV drivers";
-      field ~qualifier:DynamicRO ~ty:Bool ~in_oss_since:None "PV_drivers_up_to_date"
-	"true if the PV drivers appear to be up to date";
-
+      field ~qualifier:DynamicRO ~ty:Bool ~in_oss_since:None
+        ~lifecycle:[
+          Deprecated, rel_dundee, "Deprecated in favour of network_paths_optimized and storage_paths_optimized, and redefined in terms of them"
```
Comment (robhoes) [inline ocaml/idl/datamodel.ml]: Please add `Published, rel_rio, ""`, which is what it implicitly used to have.

File: ocaml/xapi/xapi_guest_agent.ml [hunk 2]
```diff
@@ -97,6 +98,45 @@ let networks path (list: string -> string list) =
...
+let network_paths_optimized path (list: string -> string list) (lookup: string -> string option) =
+	try
+		let _ = List.find (fun vif_id ->
+			let vif_state = lookup (extend (extend path vif_id) "state") in
+			match vif_state with
+			| Some vifstate -> vifstate <> "4"
+			| None -> true
+		) (list path) in
+		false
+	with Not_found -> true
```
Comment (robhoes) [inline ocaml/xapi/xapi_guest_agent.ml]: I find this function quite hard to follow, and it took me a while to determine that is it probably correct. This was probably because of the double (triple?) negative (`false` if we have not found not a 4, and `true` in some other cases).
I would prefer a simple fold, such as this:
```
List.fold_left (fun result vif_id ->
    let vif_state = lookup (extend (extend path vif_id) "state") in
    result && (vif_state = Some "4")
) true (list path)
```

File: ocaml/xapi/xapi_guest_agent.ml [hunk 3]
```diff
@@ -97,6 +98,45 @@ let networks path (list: string -> string list) =
...
+			let vbd_state = lookup (extend (extend path vbd_id) "state") in
+			let vbd_type = lookup (extend (extend path vbd_id) "device-type") in
+			match vbd_state, vbd_type with
+			| Some vbdstate, Some vbdtype -> vbdstate <> "4" && vbdtype <> "cdrom"
+			| None, Some vbdtype -> true
+			| Some vbdstate, None -> true
+			| None, None -> true
+		) (list path) in
+		false
+	with Not_found -> true
```
Comment (robhoes) [inline ocaml/xapi/xapi_guest_agent.ml]: As above, but even more complicated. Could this be a simple fold as well?

File: ocaml/xapi/xapi_guest_agent.ml [hunk 4]
```diff
@@ -254,8 +308,9 @@ let all (lookup: string -> string option) (list: string -> string list) ~__conte
     end;
 
 	  (* Update the 'up to date' flag afterwards *)
-	  let gmr = Db.VM_guest_metrics.get_record_internal ~__context ~self:gm in
-	  let up_to_date = Xapi_pv_driver_version.is_up_to_date (Xapi_pv_driver_version.of_guest_metrics (Some gmr)) in
+	  (* Only when both networks and disks paths are optimized, the 'up to date' flag is set to true. *)
+	  let up_to_date = Db.VM_guest_metrics.get_storage_paths_optimized ~__context ~self:gm &&
+	                   Db.VM_guest_metrics.get_network_paths_optimized ~__context ~self:gm in
 	  Db.VM_guest_metrics.set_PV_drivers_up_to_date ~__context ~self:gm ~value:up_to_date;
```
Comment (robhoes) [inline ocaml/xapi/xapi_guest_agent.ml]: Can't we just do `let up_to_date = network_paths_optimized && storage_paths_optimized` rather than getting those flags from the DB, since we've just computed them above?

Also, could we add PV_drivers_up_to_date to the cache, and only update it if changed, as done for other metrics above? We don't need a complete guest metrics record anymore to compute this field.

File: ocaml/xapi/xapi_guest_agent.ml [hunk 5]
```diff
@@ -288,6 +343,8 @@ let all (lookup: string -> string option) (list: string -> string list) ~__conte
 		  other, (* not the cached version *)
 		  memory_cached,
 		  device_id_cached,
+	      network_paths_optimized_cached,
+	      storage_paths_optimized_cached,
```
Comment (robhoes) [inline ocaml/xapi/xapi_guest_agent.ml]: Indentation looks a little off.

**Comment by robhoes** [comment]
This generally looks good to me, apart from a few suggestions above. Finally, it would be good to add a little more text to the commit message, explaining what the changes are for (there actually should have been two commit, since there are two distinct changes). I.e. 1) what are these new fields for; 2) how is the old field redefined and why?

File: ocaml/idl/datamodel.ml (hunk 1 above)
Comment (huizh) [inline ocaml/idl/datamodel.ml]: Thanks Rob. I will add it.

File: ocaml/xapi/xapi_guest_agent.ml (hunk 5 above)
Comment (huizh) [inline ocaml/xapi/xapi_guest_agent.ml]: Thanks. I will update.

File: ocaml/xapi/xapi_guest_agent.ml (hunk 4 above)
Comment (huizh) [inline ocaml/xapi/xapi_guest_agent.ml]: Yes, I agree we don't need to get flags from DB then compute pv_drivers_up_to_date. Also I will add pv_drivers_up_to_date to the cache as well.
Thanks for the suggestion.

File: ocaml/xapi/xapi_guest_agent.ml (hunk 2 above)
Comment (huizh) [inline ocaml/xapi/xapi_guest_agent.ml]: I tried to stop the compute and return false if any vifstate <> 4, so I used List.find but made the code not easy to follow.
Considering the limited count of VIF and to make code clear, I will update to use List.fold_left per your suggestion.
Thanks for your example.

File: ocaml/xapi/xapi_guest_agent.ml (hunk 3 above)
Comment (huizh) [inline ocaml/xapi/xapi_guest_agent.ml]: Yes, will update with fold too.
Thanks.

**Comment by huizh** [comment]
Hi @robhoes, thanks a lot for your comments and suggestions.
I've updated accordingly.
Would you please help review?

Thanks.

**Comment by robhoes** [comment]
@huizh This looks good now, thanks for the update. I'll merge it now.

**Comment by robhoes** [comment]
(Btw: the test build failed, but that seems to be due to infrastructure issues not related to the patch)

**Comment by huizh** [comment]
Thanks @robhoes.
